Core/MIPS/IR/IRRegCache.cpp

```diff
@@ -0,0 +1,61 @@
+#include "Core/MIPS/IR/IRRegCache.h"
+#include "Core/MIPS/IR/IRInst.h"
+
+void IRRegCache::Flush(int rd) {
+	if (rd == 0) {
+		return;
+	}
+	if (reg_[rd].isImm) {
+		ir_->WriteSetConstant(rd, reg_[rd].immVal);
+		reg_[rd].isImm = false;
+	}
+}
+
+void IRRegCache::Discard(int rd) {
+	if (rd == 0) {
+		return;
+	}
+	reg_[rd].isImm = false;
+}
+
+IRRegCache::IRRegCache(IRWriter *ir) : ir_(ir) {
+	memset(&reg_, 0, sizeof(reg_));
+	reg_[0].isImm = true;
+	ir_ = ir;
+}
+
+void IRRegCache::FlushAll() {
+	for (int i = 0; i < TOTAL_MAPPABLE_MIPSREGS; i++) {
+		//if (i < IRTEMP_0)
```

Comment:
Ah, I know that some mips will even generate dead reg stores, e.g.:

```
li a0, 5
add a0, a1, a2
```
(I don't remember digging in, maybe the cases I've seen have all been branch things...)

We could probably have a generic pass to eliminate all clobbered / unused values.

-[Unknown]